### PR TITLE
[Application] Refactor app module to fetch on-chain params

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Summary

Refactor the `application` module to remove all usage of the `NumBlocksPerSession` stand-in constant, in favor of fetching the on-chain param. This involves adding the `shared` module as a dependency.

## Issue

- #517 

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

**Documentation changes** (only if making doc changes)
- [ ] `make docusaurus_start`; only needed if you make doc changes

**Local Testing** (only if making code changes)
- [x] **Unit Tests**: `make go_develop_and_test`
- [x] **LocalNet E2E Tests**: `make test_e2e`
- See [quickstart guide](https://dev.poktroll.com/developer_guide/quickstart) for instructions

**PR Testing** (only if making code changes)
- [x] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.
    - **THIS IS VERY EXPENSIVE**, so only do it after all the reviews are complete.
    - Optionally run `make trigger_ci` if you want to re-trigger tests without any code changes
    - If tests fail, try re-running failed tests only using the GitHub UI as shown [here](https://github.com/pokt-network/poktroll/assets/1892194/607984e9-0615-4569-9452-4c730190c1d2)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [x] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

**Review by Olshansk — CHANGES_REQUESTED** [review]
Will re-review the entire PR on the next round but my main feedback is:

1. s/queryHeight/blockHeight/
2. s/GetDefaultX/GetXForDefaultParams

Lmk wyt

File: x/application/types/expected_keepers.go [hunk 1]
```diff
@@ -30,3 +31,9 @@ type BankKeeper interface {
 type GatewayKeeper interface {
 	GetGateway(ctx context.Context, addr string) (gatewaytypes.Gateway, bool)
 }
+
+// SharedKeeper defines the expected interface needed to retrieve shared information.
+type SharedKeeper interface {
+	GetParams(ctx context.Context) sharedtypes.Params
+	GetSessionEndHeight(ctx context.Context, queryHeight int64) int64
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply x/application/types/expected_keepers.go:38]: I would argue that `queryHeight` has more semantically meaning than `blockHeight`, where "block" seems redundant in my view. While the word "block" by itself is ambiguous, "height" is completely unambiguous on its own. The word "query" is then adding semantic meaning by indicating that it is a free & arbitrary parameter (as opposed to having to be a session start/end height or something).

While it's not currently the case (see #543), methods like this which depend on param values SHOULD be using the (potentially historical) value of the param **at `queryHeight`**. I think this was the original reason that "query" stuck in my head.

File: x/shared/session.go [hunk 2]
```diff
@@ -14,42 +16,85 @@ const NumBlocksPerSession = 4
...
 
-// GetSessionNumber returns the session number given the block height.
+// GetSessionNumber returns the session number given the block height given the passed
+// shared on-chain parameters.
 // Returns session number 0 if the block height is not a consensus produced block.
 // Returns session number 1 for block 1 to block NumBlocksPerSession - 1 (inclusive).
 // i.e. If NubBlocksPerSession == 4, session == 1 for [1, 4], session == 2 for [5, 8], etc.
-func GetSessionNumber(blockHeight int64) int64 {
-	if blockHeight <= 0 {
+func GetSessionNumber(sharedParams *sharedtypes.Params, queryHeight int64) int64 {
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply x/shared/session.go:90]: Consolidating discussion in https://github.com/pokt-network/poktroll/pull/555#discussion_r1617787650

File: x/shared/session.go [hunk 3]
```diff
@@ -14,42 +16,87 @@ const NumBlocksPerSession = 4
...
 	// TODO_BLOCKER(#543): If the num_blocks_per_session param has ever been changed,
 	// this function may cause unexpected behavior.
-	return blockHeight - ((blockHeight - 1) % NumBlocksPerSession)
+	return queryHeight - ((queryHeight - 1) % numBlocksPerSession)
+}
+
+// GetSessionEndHeightWithDefaultParams returns the block height at which the session
+// containing queryHeight ends, given the default shared on-chain parameters.
+// Returns 0 if the block height is not a consensus produced block.
+// Example: If NumBlocksPerSession == 4, sessions end at blocks 4, 8, 11, etc.
```
Comment (Olshansk) [inline x/shared/session.go:50]:
```suggestion
// See GetSessionEndHeight for more details.
```

File: x/shared/session.go [hunk 4]
```diff
@@ -14,42 +16,87 @@ const NumBlocksPerSession = 4
 // governance parameter is implemented.
 const SessionGracePeriodBlocks = 4
 
-// GetSessionStartBlockHeight returns the block height at which the session starts
+// GetSessionStartHeightWithDefaultParams returns the block height at which the
+// session containing queryHeight starts, given the default shared on-chain
+// parameters.
+// Returns 0 if the block height is not a consensus produced block.
+// Example: If NumBlocksPerSession == 4, sessions start at blocks 1, 5, 9, etc.
```
Comment (Olshansk) [inline x/shared/session.go:23]:
```suggestion
// See GetSessionStartHeight for more details.
```

File: x/shared/session.go [hunk 5]
```diff
@@ -14,42 +16,87 @@ const NumBlocksPerSession = 4
...
+	numBlocksPerSession := int64(sharedParams.GetNumBlocksPerSession())
+
+	return GetSessionStartHeight(sharedParams, queryHeight) + numBlocksPerSession - 1
+}
+
+// GetSessionNumberWithDefaultParams returns the session number of the session
+// containing queryHeight, given the default on-chain shared parameters.
+// Returns session number 0 if the block height is not a consensus produced block.
+// Returns session number 1 for block 1 to block NumBlocksPerSession - 1 (inclusive).
+// i.e. If NubBlocksPerSession == 4, session == 1 for [1, 4], session == 2 for [5, 8], etc.
```
Comment (Olshansk) [inline x/shared/session.go:76]:
```suggestion
// See GetSessionNumber for more details.
```

**Review by Olshansk — APPROVED** [review]
Optional comment updates

**Comment by bryanchriswhite** [comment]
:rotating_light: Do not delete branch until #556's base has changed! :rotating_light:

**Pull request merged**